services/subscriber/service.go

```diff
@@ -0,0 +1,204 @@
+package subscriber
+
+import (
+	"expvar"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/influxdb/influxdb"
+	"github.com/influxdb/influxdb/cluster"
+	"github.com/influxdb/influxdb/meta"
+)
+
+// Statistics for the CQ service.
+const (
+	statPointsWritten = "points_written"
+)
+
+type PointsWriter interface {
+	WritePoints(p *cluster.WritePointsRequest) error
+}
+
+// unique set that identifies a given subscription
+type subEntry struct {
+	db   string
+	rp   string
+	name string
+}
+
+type Service struct {
+	subs      map[subEntry]PointsWriter
+	MetaStore interface {
+		Databases() ([]meta.DatabaseInfo, error)
+	}
+	NewPointsWriter func(u url.URL) (PointsWriter, error)
+	Logger          *log.Logger
+	statMap         *expvar.Map
+	ticker          *time.Ticker
+}
+
+func NewService() *Service {
+	return &Service{
+		subs:            make(map[subEntry]PointsWriter),
+		NewPointsWriter: newPointsWriter,
+		Logger:          log.New(os.Stderr, "[subscriber] ", log.LstdFlags),
+		statMap:         influxdb.NewStatistics("subscriber", "subscriber", nil),
+	}
+}
+
+func (s *Service) Open() error {
+	if s.MetaStore == nil {
+		panic("no meta store")
+	}
+	s.Update()
+	go s.pollForMetaUpdates()
+
+	return nil
+}
+
+func (s *Service) Close() error {
+	if s.ticker != nil {
+		s.ticker.Stop()
+	}
+	return nil
+}
+
+func (s *Service) pollForMetaUpdates() {
+	s.ticker = time.NewTicker(time.Minute)
+	for range s.ticker.C {
+		s.Update()
+	}
+}
+
+// start new and stop deleted subscriptions.
+func (s *Service) Update() error {
+	dbis, err := s.MetaStore.Databases()
+	if err != nil {
+		return err
+	}
+	allEntries := make(map[subEntry]bool, 0)
+	// Add in new subscriptions
+	for _, dbi := range dbis {
+		for _, rpi := range dbi.RetentionPolicies {
+			for _, si := range rpi.Subscriptions {
+				se := subEntry{
+					db:   dbi.Name,
+					rp:   rpi.Name,
+					name: si.Name,
+				}
+				allEntries[se] = true
+				if _, ok := s.subs[se]; ok {
+					continue
+				}
+				var bm BalanceMode
+				switch si.Mode {
+				case "ALL":
+					bm = ALL
+				case "ANY":
+					bm = ANY
+				default:
+					return fmt.Errorf("unknown balance mode %q", si.Mode)
+				}
+				writers := make([]PointsWriter, len(si.Destinations))
+				statMaps := make([]*expvar.Map, len(writers))
+				for i, dest := range si.Destinations {
+					u, err := url.Parse(dest)
+					if err != nil {
+						return err
+					}
+					w, err := s.NewPointsWriter(*u)
+					if err != nil {
+						return err
+					}
+					writers[i] = w
+					tags := map[string]string{"database": se.db, "retention_policy": se.rp, "name": se.name, "mode": si.Mode}
+					statMaps[i] = influxdb.NewStatistics(dest, "subscriber", tags)
```

Comment:
The best to be *sure* you're happy with the stats is fire up the system,
generate some activity and the execute the query 'SHOW STATS' at the CLI.
How does it look?

On Fri, Oct 9, 2015 at 1:04 PM, Nathaniel Cook <notifications@github.com>
wrote:

> In services/subscriber/service.go
> <https://github.com/influxdb/influxdb/pull/4375#discussion_r41672288>:
>
> > +					return fmt.Errorf("unknown balance mode %q", si.Mode)
> > +				}
> > +				writers := make([]PointsWriter, len(si.Destinations))
> > +				statMaps := make([]*expvar.Map, len(writers))
> > +				for i, dest := range si.Destinations {
> > +					u, err := url.Parse(dest)
> > +					if err != nil {
> > +						return err
> > +					}
> > +					w, err := s.NewPointsWriter(*u)
> > +					if err != nil {
> > +						return err
> > +					}
> > +					writers[i] = w
> > +					tags := map[string]string{"database": se.db, "retention_policy": se.rp, "name": se.name, "mode": si.Mode}
> > +					statMaps[i] = influxdb.NewStatistics(dest, "subscriber", tags)
>
> Ok, so key is global to all statMaps, I updated it to be unique now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdb/influxdb/pull/4375/files#r41672288>.
>